Test: Fuzz message bus

sentientwaffle opened the pull request:
The fuzzer is capable of finding one of the two known crashes on main, plus one extra.

Additionally, it can find both of the already-known crashes on https://github.com/tigerbeetle/tigerbeetle/pull/3245 (https://github.com/tigerbeetle/tigerbeetle/commit/db4546aef5be3b0737db4992b45126c3f069d8e2), as well as a couple others.

~Caveat: This fuzzer is currently only _mostly_ deterministic. It uses real TCP for IO. However, in practice reusing the same seed should give the same results anyway.~ Fuzzer has been rewritten to be deterministic, using simulated IO!

sentientwaffle committed:
Test: Add MessageBus fuzzer

The fuzzer is capable of finding one of the two known crashes on main, plus one other.

Additionally, it can find both of the already-known crashes on https://github.com/tigerbeetle/tigerbeetle/pull/3245 (db4546aef5be3b0737db4992b45126c3f069d8e2), as well as a couple others.

Caveat: This fuzzer is currently only _mostly_ deterministic. It uses real TCP for IO. However, in practice reusing the same seed should give the same results anyway.

sentientwaffle committed:
Test/MessageBus: Avoid port collisions

sentientwaffle committed:
CFO: Add message_bus fuzzer

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,172 @@
+const fuzz = @import("testing/fuzz.zig");
+const ratio = stdx.PRNG.ratio;
+
+/// NOTE: In practice this fuzzer should be *mostly* deterministic, but it is using real network IO.
```
Ha!

I think it actually is better this way, because, well, we can poke relevant parts of IO as well. 


_But_ the fact that we depend on global port numbers here really rubs me the wrong way. But then we have unsare... 

Yeah, I think I have strong argument for doing it this way, and that is that we want to check just how good our `maybe_unsare_and_relaunch` is!

matklad replied:
Though, if I had infinite time, I _think_ I'd write "fake IO" for message bus? To fault inject better, and to make sure that there _is_ a reasonable interface between the bus and IO

sentientwaffle replied:
Hmm, I may not have a choice -- it appears that message bus fuzzer gets EPERM when trying to unshare when it is run from within the cfo supervisor's unshare. :disappointed:

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,172 @@
+        // Get our own network namespace so that multiple instances of the fuzzer can use the same
+        // ports without conflicting.
+        try stdx.unshare.maybe_unshare_and_relaunch(gpa, .{ .pid = false, .network = true });
+    }
```
Let's maybe put it at the first line of main? Before we init PRNG.

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,172 @@
+        for (message_buses, ios, 0..) |*message_bus, *io, i| {
+            if (prng.chance(tick_message_bus_probability)) {
+                message_bus.tick();
+            }
```
Nice!

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,172 @@
+const vsr = @import("vsr.zig");
+const constants = vsr.constants;
+const stdx = vsr.stdx;
+const MessageBus = vsr.message_bus.MessageBusReplica;
```
Fuzzing client's bus is also important, maybe even moreso: we know that TigerBeetle can recover from crashes of individual replicas fine, but crashing our clients applications would be impolite. 

🤔 I wonder if we need to get rid of client/replica bus distinction in the _type_? We compile in both client and replica versions (due to benchmark), and it feels wasteful!

matklad reviewed (commented):
🎉

sentientwaffle committed:
Test/MessageBus: Don't unshare

unshare from within the cfo's own unshare doesn't work due to EPERM.

sentientwaffle committed:
IO: Export ListenOptions

sentientwaffle committed:
Test/MessageBus: Fuzz with deterministic IO

Motivations:
- no nested unshare when run by cfo supervisor
- cover rare paths and errors
- deterministic
- much faster

sentientwaffle committed:
Test/MessageBus: Simulate network corruption

sentientwaffle committed:
Test/MessageBus: Inject junk data

sentientwaffle committed:
Test/MessageBus: Implement send_now()

sentientwaffle committed:
Test/MessageBus: Check pending send()/recv()

Inspired by this note from the io_uring man page:

> Concretely, for operations where strict ordering is
> required, such as for sends and receives on a stream-
> oriented TCP socket, it is generally unsafe to have more
> than one outstanding send, or more than one outstanding
> receive (the two directions are independent) on a given
> socket at a time, as the kernel may reorder their execution
> if poll arming or other background kernel activities are
> involved.

In message bus' case it isn't actually _unsafe_ since we tolerate message reordering, but it is certainly _unexpected_, so we can assert its absence.

sentientwaffle committed:
Test/MessageBus: Simulate close() errors

sentientwaffle committed:
Test/MessageBus: Simulate send()/recv() errors

sentientwaffle committed:
Test/MessageBus: Fix socket leak check

sentientwaffle committed:
Stdx/PRNG: Refactor error_uniform()

sentientwaffle committed:
Test/MessageBus: Refactor

sentientwaffle committed:
Test/MessageBus: Test shutdown() errors

sentientwaffle committed:
Test/MessageBus: Test accept()/connect() errors

sentientwaffle committed:
Test/MessageBus: Cleanup

sentientwaffle committed:
Test/MessageBus: Verify message delivery

sentientwaffle committed:
Test/MessageBus: Test MessageBusClient too

sentientwaffle committed:
Test/MessageBus: Add back misdirection faults for peer misidentification

sentientwaffle committed:
Test/MessageBus: Misc

sentientwaffle commented:
Alright, this is ready for review again!

It has been almost completely rewritten, to use simulated IO instead of real IO.

Additionally:
- Since IO is simulated, simulate errors too.
- Check that nodes (eventually) receive the messages they expect to receive. (How we ensure peer identification and misidentification has been changed to accommodate this).
- Test both `MessageBusReplica` and `MessageBusClient`.
- Test network corruption.

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+                const result_accept: AcceptError!socket_t = local_fd;
+                const result_connect: ConnectError!void = {};
+
+                completion.callback(
+                    completion.context,
+                    completion,
+                    &result_accept,
+                );
+                remote_completion.callback(
+                    remote_completion.context,
+                    remote_completion,
+                    &result_connect,
+                );
```
neeeeat!

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+    connections_closing: u32 = 0,
+    events: EventQueue,
+    ticks: u64 = 0,
+    fd: socket_t = 1,
```
maybe call it `fd_next`? It's not an fd, its a counter we used to create one.

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+        .close_error_probability = ratio(prng.int_inclusive(u64, 2), 10),
+        .shutdown_error_probability = ratio(prng.int_inclusive(u64, 2), 10),
+        .accept_error_probability = ratio(prng.int_inclusive(u64, 2), 10),
+        .connect_error_probability = ratio(prng.int_inclusive(u64, 2), 10),
```
Unrelated, but I wish we had some nicer API to create a random ratio out of tree numbers, but I just don't seem to find something I like `prng.ratio(0, 2, 10)` is too cryptic.

sentientwaffle replied:
:thinking: Maybe `prng.ratio(.{0, 2}, 10)`?

matklad replied:
Yeah, I thought about that, but it doesn't really look all that nice.

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+    }
+
+    fn tick_instant(io: *const IO) stdx.Instant {
+        return .{ .ns = io.ticks * constants.tick_ms * std.time.ns_per_ms };
```
Could also do `return .ms(io.ticks * constants.ticks_ms)`. I considered and explicitly rejected the idea of adding `.ticks` constructor to `Duration` --- I think that would be poor layering.

sentientwaffle replied:
`Instant` doesn't have `.ms()` atm, and I'm not convinced we should add it, since constructing an Instant that way (outside of testing) is wrong.

matklad replied:
Ah, right, that's instant, not Duration! Yeah, instant must not have any convenience constructors, as construction an Instant by hand is almost always a bug (but fuzzers are exception!)

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+        io.events.add(.{
+            .completion = completion,
+            .ready_at = io.tick_instant().add(.{ .ns = delay_ns }),
+        }) catch unreachable;
```
Maybe `@panic("OOM")` would be a bit more readable here.

matklad replied:
Hm, if we use `@panic("OOM")` here, maybe we should consistently use it in the whole file? E.g., 

```zig
io.servers.putNoClobber(io.gpa, fd, .{ .address = address }) catch unreachable;
```

basically, my thinking here is that every time I see `catch unreachable`, I need to do some mental calculation to figure out why is this unreachable. If I see `@panic("OOM")`, I clearly understand that there _isn't_ some internal invariant I need to uphold, and rather that there's just an allocation. Of course, that "trusts" the writer that its indeed oom and not something else, but that can be checked once. 

In particular, in the above example there are two possibilities: either we allocate, _or_ `putNoClobber` returns an `error.WouldClobber`, and I need to refresh my knowledge of hash-map API to find out which is which. With `@panic("OOM")`, the intention would be more clear!

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+                        try sender.sending.append(gpa, io.prng.int(u8));
+                    } else {
+                        const corrupt_byte = io.prng.index(send_buffer);
+                        const corrupt_bit = io.prng.int_inclusive(u3, @bitSizeOf(u8) - 1);
```
I guess we can introduce `prng.random_bit(u8)` here? We use that in testing/storage.zig and routing.zig

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,915 @@
+        comptime callback: fn (Context, *Completion, TimeoutError!void) void,
+        completion: *Completion,
+        nanoseconds: u63,
+    ) void {
```
Not really related, but we should split `IO.timeout` into `IO.next_tick` and `timeout` proper, where `timeout` asserts that the delay is strictly positive!

matklad reviewed (commented):
On the cursory look, looks awesome! Don't have enough brain left today to really dig into though!

sentientwaffle committed:
Stdx/PRNG: prng.bit()

sentientwaffle committed:
Test/MessageBus: Code review

matklad commented on src/stdx/prng.zig:
```diff
@@ -375,6 +375,23 @@ test boolean {
 
+/// Returns the index of a randomly-chosen bit within a Word.
+pub fn bit(prng: *PRNG, comptime Word: type) std.math.Log2Int(Word) {
+    comptime assert(@typeInfo(Word) == .int);
```
Also assert that it is unsigned.

matklad replied:
Ah, wait, no, I had a slightly different function in mind here!

```zig
pub fn bit(prng: *PRNG, comptime Word: type) Word {
    comptime assert(@typeInfo(Word) == .int);
    return @as(Word, 1) << prng.int_inclusive(std.math.Log2Int(Word), @bitSizeOf(Word) - 1);
}
```

So that the call site is just 

```zig
sender.sending.items[sender.sending.items.len - corrupt_byte - 1] ^= io.prng.bit(u8);
```

matklad replied:
Though, if naming/intention is confusing, maybe better not to have it? Though, it feels like it does capture a rather crisp concept!

sentientwaffle committed:
Stdx/PRNG: prng.bit() returns a word with one bit set

sentientwaffle committed:
Test/MessageBus: Code review

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,914 @@
+        return address;
+    }
+
+    pub fn close_socket(io: *IO, socket: socket_t) void {
```
TIL we have both (async) `fn close` and (sync) `fn close_socket`! That feels suspicious, but not something we should be immediately fixing

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,914 @@
+    pub fn open_socket_tcp(io: *IO, _: u32, _: RealIO.TCPOptions) !socket_t {
+        const socket = io.fd_next;
+        io.fd_next += 1;
+        return @intCast(socket);
```
Let's assert that if any socket that is created, is also closed. One thing I worry most with message bus is that we have an fd-leak in the reconnect loop, and might run out of file descriptors!

Alternatively, we can assert that for any node, the number of simultaneously open connections is bounded.

matklad replied:
Note that we also "auto close" on send error: 

```
                    // If there is a pending recv(), we need to make sure that it fails (rather than
                    // potentially stalling and preventing MessageBus from calling shutdown/close).
                    sender.closed = true;
```

but we expect message bus to still explicitly close the socket. So perhaps the `closed` needs to be split into "closed externally by the message bus" and "closed internally due to error".

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,914 @@
+
+                    // If there is a pending recv(), we need to make sure that it fails (rather than
+                    // potentially stalling and preventing MessageBus from calling shutdown/close).
+                    sender.closed = true;
+                    return .done;
+                }
+
+                if (sender.shutdown_send) {
+                    const result: SendError!usize = error.BrokenPipe;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
```
Hm, can we do it probabilistically? I am not entirely sure, but I think if you call `send`, and then `shutdown`, the `send` might still complete after shutdown?

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,914 @@
+        completion: *Completion,
+        fd: fd_t,
+    ) void {
+        completion.* = .{
```
Assert that there are no pending IO operations on socket here. `maybe_close` has some tricky logic to only call close without outstanding IO, and would be great to verify that that works correctly. 

Really, looking at this closer, I can't believe that we managed to write message bus without a fuzzer, there are _so_ many tricky invariants here, thanks for fixing this  once and for all with a fuzzer!

matklad replied:
Ah, maybe we should also set `.closed` state immediately here, and not in completion?

sentientwaffle replied:
> Ah, maybe we should also set .closed state immediately here, and not in completion?

:thinking: I don't follow -- isn't setting it asynchronously closer to what "real" IO would do? Though I guess it doesn't strictly matter since if we call `close()` there shouldn't be any more operations on that socket anyhow...

matklad replied:
If we set as logically closed synchronously, we'll get stricter asserts, as

```
io.close(socket, callback1); 
io.send(socket, callback2);
```


will fail.

matklad commented on src/message_bus_fuzz.zig:
```diff
@@ -0,0 +1,914 @@
+                    const result: RecvError!usize = 0;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
```
Similar here, I'd expect that it's possible that recv initiated before shutdown can complete after. Actually, I'd expect that to be the normal case? Shutdown flushes everything already submitted, and then closes, rather than abruptly cutting

sentientwaffle replied:
Ooh, good catch! I was going off of the man page which says:

> If `how` is `SHUT_RD`, further receptions will be  disallowed.

...but I think that is misleading, and you can still receive anything that had already been buffered!

matklad replied:
I think another guiding principle is that we shouldn't cut it too close to the manpage, if we can impose _less_ restrictions on semantics, it might better to do that!
